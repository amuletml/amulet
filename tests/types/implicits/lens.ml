external val tostring : 'a -> string = "tostring"
external val print : string -> unit = "print"

type identity 'a = Identity of 'a
let runIdentity (Identity x) = x

type const 'a 'b = Const of 'a
let getConst (Const x) = x

type functor 'f = Functor of (forall 'a 'b. ('a -> 'b) -> 'f 'a -> 'f 'b)

let implicit functor_identity =
  Functor (fun f (Identity x) -> Identity (f x))

let implicit functor_const =
  Functor (fun _ (Const x) -> Const x)

let f & g = fun x -> f (g x)
let x |> f = f x

let (<$>) ?(Functor f) = f

let fst ?f k (a, b) = (,b) <$> k a

let snd ?f k (a, b) = (a,) <$> k b

let l .~ b = runIdentity & l (fun _ -> Identity b)

let s ^. l = getConst (l Const s)

let print_pair (x, y) = print ("(" ^ tostring x ^ ", " ^ tostring y ^ ")")

let main =
  let p = (1, "foo")
  print_pair (p |> snd .~ 2)
  print (tostring (p ^. fst))
