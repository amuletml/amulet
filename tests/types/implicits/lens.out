tostring : {'a : type}. 'a -> string
print : string -> unit
identity : type -> type
Identity : {'a : type}. 'a -> identity 'a
runIdentity : {'cf : type}. identity 'cf -> 'cf
const : {'cx : type}. type -> 'cx -> type
Const : {'a : type}. {'b : 'dk}. 'a -> const 'a 'b
getConst : {'du : type}. {'b : 'ei}. const 'du 'b -> 'du
functor : (type -> type) -> type
Functor : {'f : type -> type}. ({'a : type}. {'b : type}. ('a -> 'b) -> 'f 'a -> 'f 'b) -> functor 'f
functor_identity : functor identity
functor_const : {'a : type}. functor (const 'a)
& : {'jj : type}. {'jk : type}. {'jx : type}. ('jx -> 'jk) -> ('jj -> 'jx) -> 'jj -> 'jk
|> : {'kr : type}. {'kt : type}. 'kr -> ('kr -> 'kt) -> 'kt
<$> : {'f : type -> type}. {'a : type}. {'b : type}. functor 'f => ('a -> 'b) -> 'f 'a -> 'f 'b
fst : {'ns : type}. {'f : type -> type}. {'of : type}. {'oi : type}. functor 'f => ('oi -> 'f 'of) -> ('oi * 'ns) -> 'f ('of * 'ns)
snd : {'qb : type}. {'f : type -> type}. {'qp : type}. {'qs : type}. functor 'f => ('qs -> 'f 'qp) -> ('qb * 'qs) -> 'f ('qb * 'qp)
.~ : {'rw : type}. {'jj : type}. {'jk : type}. {'sw : type}. (('sw -> identity 'rw) -> 'jj -> identity 'jk) -> 'rw -> 'jj -> 'jk
^. : {'tv : type}. {'tx : type}. {'b : 'va}. {'a : type}. {'b : 'vc}. 'tv -> (('a -> const 'a 'b) -> 'tv -> const 'tx 'b) -> 'tx
print_pair : {'a : type}. {'a : type}. ('a * 'a) -> unit
main : unit
